pydantic/main.py

```diff
@@ -75,6 +77,42 @@ def __pydantic_post_init__(self_: Any, **kwargs: Any) -> None:
             validator_functions = _validation_functions.ValidationFunctions(bases)
             namespace['__pydantic_validator_functions__'] = validator_functions
 
+           # Errors for misuse of discriminated unions (check test_discriminated_unions)
+            for name, value in namespace['__annotations__'].items():
```

Comment:
this logic should be in 

https://github.com/pydantic/pydantic/blob/55fd51fbdfea44cc16b41bdda8e102c85aefcbfc/pydantic/_internal/_model_construction.py#L167

Also we shouldn't be looking at `namespace['__annotations__']` directly anymore, but instead using `get_type_hints` as done in the above function.